Autopilot enterprise docs (#15589)
hashicorp/vault: website/content/docs/enterprise/automated-upgrades.mdx

---
layout: docs
page_title: Vault Enterprise Automated Upgrades
description: |-
  Vault Enterprise can upgrade itself automatically.
---

# Automated Upgrades

> -> **Note**: This feature requires [Vault Enterprise](https://www.hashicorp.com/products/vault/)

Reviewer: ```suggestion
~> **Note**: Automated Upgrades require [Vault Enterprise](https://www.hashicorp.com/products/vault/) to be 
```

> configured to use Integrated Storage.

Reviewer: ```suggestion
```

> Vault Enterprise Automated Upgrades allows operators to upgrade the version of Vault running in

Reviewer: ```suggestion
Vault Enterprise Automated Upgrades allows operators to upgrade the Vault version currently running in cluster automatically. 
```

> a cluster in an automated fashion. There are a few different ways to make this upgrade happen,

Reviewer: ```suggestion
 There are a few different ways to make this upgrade happen,
```

> and control which versions are being upgraded to. By default, with no additional configuration,

Reviewer: ```suggestion
and control which versions are being upgraded to. With no additional configuration,
```

Vault will check the version of Vault that each node in the cluster is running. If a blue/green

> style deploy is desired, Vault can be made to use a version of your choosing, regardless of

Reviewer: ```suggestion
style deployment is desired, Vault uses the version of your choosing, regardless of
```

> what version of Vault is currently running.

Reviewer: ```suggestion
which version of Vault is currently running.
```

## Configuration
A new key can be added to Vault's `storage` configuration stanza: `autopilot_upgrade_version`.
The value for this key is a [SemVer](https://semver.org/) compatible version string of your choosing.

> If this is present, it will override the actual version of Vault that is running for the purposes of

Reviewer: ```suggestion
When  version string is present, it will override the current version of Vault that is running to upgrade automatically.
```

> automated upgrades.

Reviewer: ```suggestion

```

## Mechanics

> Regardless of whether you choose to use Vault's built-in version, or a version of your choosing, the

Reviewer: ```suggestion
Whether you choose to use Vault's built-in version, or a version of your own, the
```

> mechanics for performing an automated upgrade are the same. Once you have a Vault cluster running,

Reviewer: ```suggestion
mechanics for performing an automatic upgrades remain the same. Once you have a Vault cluster running,
```

Reviewer: Under the Mechanics section, this is my suggested rewrite. Please feel fee to ignore if the rewrite changed the intent of this paragraph in any way.

**Suggested changes**: 
Whether you choose to use Vault's built-in version or a version of your own, the mechanics for performing automatic upgrades remain the same.

When a Vault cluster is running and new nodes containing an updated Vault version join the cluster, the Autopilot subsystem within Vault will promote the new version nodes to voters when the number of nodes running the latest Vault version equals or exceeds the number of pre-existing nodes. Leadership transfers from the prior leader to a randomly selected node running the newest Vault version and waits for the user to remove the previous nodes from the cluster. Vault then demotes the previous version's nodes to non-voters.

Author: It does change it slightly, but I like a lot of your wording better than mine. I will use your paragraph as much as possible, with a few slight tweaks.

start and join a number of new nodes to the cluster, running the updated version. Once the number of
nodes running the new version equals or exceeds the number of nodes running the old version, the
Autopilot subsystem within Vault will promote the new version nodes to become voters, demote the old
version nodes to become non-voters, transfer leadership from the old leader to a randomly selected
node running the new version, and wait for you to remove the old version nodes from the cluster.

Below is a flowchart depicting Autopilot's automated upgrade state machine.

![Automated Upgrade State Machine](/img/autopilot-automated-upgrade.png)

The status of the automated upgrade can be monitored by consulting the [Autopilot state API endpoint](/api-docs/system/storage/raftautopilot#get-cluster-state).

## Examples
### Using Vault's built-in version

> This is the easiest to perform an automated upgrade. No configuration is needed. Automated upgrades

Reviewer: ```suggestion
This is the easiest method to perform an automated upgrade; no configuration is needed. Automated upgrades
```

> are enabled by default. Following the procedure outlined above in the [Mechanics](#mechanics) section

Reviewer: Following the procedure outlined above... this sentence is redundant so you may want to consider removing.
